src/k2/common/MapWithExpiry.h

```diff
@@ -0,0 +1,160 @@
+    }
+
+    // TODO: Implement using boost iterator helpers to reduce boilerplate code.
+    iterator end() {return elems.end();}
+    iterator begin() { return elems.begin();}
+    iterator find(const KeyT &key) {return  elems.find(key);}
+    ValT& at(const KeyT &key) {return elems.at(key).getValue();}
+    auto size() {return elems.size();}
+    void unlink(iterator iter) {
+        tsOrderedList.erase(tsOrderedList.iterator_to(iter->second));
+    }
+    auto extract(iterator iter) {
+        unlink(iter);
+        // Remove from map
+        return elems.extract(iter);
+    }
+    auto erase(const KeyT &key) {
+        iterator iter = elems.find(key);
+        unlink(iter);
+        return elems.erase(iter);
+    }
+
+    void resetTs(iterator iter, TimePoint ts) {
+        // new TS needs to be >= current maximum TS
+        if (ts < tsOrderedList.back().ts) {
+            assert(false);
+            throw std::invalid_argument("too small ts");
+        }
+        unlink(iter);
+        iter->second.ts = ts;
+        tsOrderedList.push_back(iter->second);
+    }
+
+    iterator getFirst() {
+        if (tsOrderedList.empty())
+            return elems.end();
+        return find(tsOrderedList.front().key);
+    }
+
+    TsOrderedList tsOrderedList;
+    std::unordered_map<KeyT, MapElem> elems;
+};
+
+template <class KeyT, class ValT, class ClockT=Clock>
+class MapWithExpiry {
+public:
+    template <typename Func>
+    void start(Duration timeout, Func&& func) {
+        _timeout = timeout;
+        _expiryTimer.setCallback([this, func = std::move(func)] {
+            return seastar::do_until(
+                [this] {
+                    auto iter = _elems.getFirst();
+                    return (iter == _elems.end() || iter->second.timestamp() > ClockT::now());
+                },
+                [this, func] {
+                    auto iter = _elems.getFirst();
+                    // First Remove element from map before calling callback
+                    auto node = _elems.extract(iter);
+                    if (!node) return seastar::make_ready_future();
+                    return func(node.key(), node.mapped().getValue());
+                });
+
+        });
+        _expiryTimer.armPeriodic(_timeout/2);
+     }
+
+    seastar::future<> stop() {
+        return _expiryTimer.stop()
+            .then([this] {
+                std::vector<seastar::future<>> bgFuts;
+                for(auto iter = _elems.begin(); iter != _elems.end(); iter++) {
+                    _elems.unlink(iter);
+                    bgFuts.push_back(iter->second.getValue().cleanup().discard_result());
+                }
+                return seastar::when_all_succeed(bgFuts.begin(), bgFuts.end());
+            });
+    }
+
+    typedef typename TimestampOrderedMap<KeyT, ValT>::iterator iterator;
+
+    // Check if key is present, with optionally push back expiry if present
+    iterator find(const KeyT &key, bool reorder) {
```

Comment:
Maybe a separate find_and_reorder function instead of the bool? Or an enum instead of the bool? It makes it a bit hard to read the usage of this function as it is.

Reply:
Fixed.